src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1104 @@
+                if (!IsSupportedLocation(aSelectedLocation))
+                {
+                    exitResponse(SelectLocationsStatus::kUnsupportedLocation, ""_span);
+                    return;
+                }
+
+                // Checking for duplicate locations.
+                uint32_t j         = 0;
+                auto jLocationIter = req.newLocations.Value().begin();
+                while (j < i)
+                {
+                    jLocationIter.Next();
+                    if (jLocationIter.GetValue() == aSelectedLocation)
+                    {
+                        exitResponse(SelectLocationsStatus::kDuplicatedLocations, ""_span);
+                        return;
+                    }
+                    j += 1;
+                }
+
+                // check to see if parameter list and attribute still match
+                if (matchesCurrentSelectedLocations)
+                {
+                    if (!mDelegate->GetSelectedLocationByIndex(listIndex, oldSelectedLocation) ||
+                        (aSelectedLocation != oldSelectedLocation))
+                    {
+                        matchesCurrentSelectedLocations = false;
+                    }
+                }
+
+                i += 1;
+            }
+
+            // after iterating with Next through DecodableType - check for failure
+            if (CHIP_NO_ERROR != iLocationIter.GetStatus())
+            {
+                exitResponse(SelectLocationsStatus::kInvalidSet, "decoding of newLocations parameter failed"_span);
+                return;
+            }
+        }
+    }
+
+    // If the NewLocations field is the same as the value of the SelectedLocations attribute
+    // the SelectLocationsResponse command SHALL have the Status field set to Success and
+    // the StatusText field MAY be supplied with a human-readable string or include an empty string.
+    if (matchesCurrentSelectedLocations)
+    {
+        exitResponse(SelectLocationsStatus::kSuccess, ""_span);
+        return;
+    }
+
+    char delegateStatusBuffer[kMaxSizeStatusText];
+    MutableCharSpan delegateStatusText(delegateStatusBuffer);
+
+    // If the current state of the device doesn't allow for the locations to be selected,
+    // the SelectLocationsResponse command SHALL have the Status field set to InvalidInMode.
+    // if the Status field is set to InvalidInMode, the StatusText field SHOULD indicate why the request is not allowed,
+    // given the current mode of the device, which may involve other clusters.
+    // (note - locationStatusText to be filled out by delegated function for if return value is false)
+    if (!mDelegate->IsSetSelectedLocationsAllowed(delegateStatusText))
+    {
+        exitResponse(SelectLocationsStatus::kInvalidInMode, delegateStatusText);
+        return;
+    }
+
+    // Set to empty string in case the delegate accidentally modified this string.
+    chip::CopyCharSpanToMutableCharSpan(""_span, delegateStatusText);
```

Comment:
After this, `delegateStatusText` is 0-length, so delegates can't put anything into it.

The right thing to do is to reset it to point to the whole `delegateStatusBuffer`.  Or better yet, have separate scoped `delegateStatusText` instances for all the uses.

Also: this was clearly not tested....

Reply:
I can replace this with `delegateStatusText = MutableCharSpan(delegateStatusBuffer);`. However, this dose not solve the thing I was trying to solve with this line (which now I realise created another problem). This is that if the delegate chooses not to update `delegateStatusText`, we will be printing random things from memory. Ideally, if the delegate chooses not to update this, the result is an empty string. I don't think this is possible and therefore we need to make the consequences clear in the delegate method documentation.

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/5cffc6876622679c1f0e1924c31cd16f28fa3db0).